File: scrape_kfw_zinsen.php
<?php
/**
 * KfW 300 und Interhyp Zinsen Scraper (PHP Version)
 * Scrapet aktuelle Zinssätze und speichert sie in zinsen-kfw300.json
 */

// Fehlerausgabe für Debugging (bei Problemen)
// error_reporting(E_ALL);
// ini_set('display_errors', 1);

/**
 * HTML von einer URL laden
 */
function fetch_html($url) {
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36');
    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
    $html = curl_exec($ch);
    $error = curl_error($ch);
    curl_close($ch);
    
    if ($error) {
        echo "cURL Fehler: $error\n";
        return false;
    }
    
    return $html;
}

/**
 * KfW-Zinssätze von der offiziellen API abrufen
 */
function scrape_kfw_zinsen_api() {
    echo "Rufe KfW-API ab...\n";
    
    $api_url = 'https://www.kfw.de/zinsen/soll.rates';
    
    // API aufrufen
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $api_url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36');
    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
    $json_response = curl_exec($ch);
    $error = curl_error($ch);
    curl_close($ch);
    
    if ($error) {
        echo "cURL Fehler: $error\n";
        return null;
    }
    
    if (!$json_response) {
        echo "Keine Antwort von der API\n";
        return null;
    }
    
    // JSON dekodieren
    $data = json_decode($json_response, true);
    
    if ($data === null) {
        echo "JSON-Dekodierung fehlgeschlagen\n";
        echo "Response (erste 500 Zeichen): " . substr($json_response, 0, 500) . "\n";
        return null;
    }
    
    // Die API könnte verschiedene Strukturen zurückgeben:
    // Variante 1: {"kfw": [...]}
    // Variante 2: Direct array [...]
    // Variante 3: Verschachtelt z.B. {"data": {"kfw": [...]}}
    
    $kfw_programs = null;
    
    if (isset($data['kfw']) && is_array($data['kfw'])) {
        $kfw_programs = $data['kfw'];
        echo "✓ API-Struktur: kfw-Array gefunden\n";
    } elseif (is_array($data) && isset($data[0]) && isset($data[0]['programm'])) {
        // Direktes Array von Programmen
        $kfw_programs = $data;
        echo "✓ API-Struktur: direktes Programm-Array\n";
    } elseif (isset($data['data']) && isset($data['data']['kfw'])) {
        $kfw_programs = $data['data']['kfw'];
        echo "✓ API-Struktur: verschachteltes kfw-Array\n";
    } else {
        echo "⚠ API-Response hat unerwartetes Format\n";
        echo "Verfügbare Keys: " . implode(', ', array_keys($data)) . "\n";
        echo "Response (erste 300 Zeichen): " . substr($json_response, 0, 300) . "\n";
        return null;
    }
    
    // Filtere KfW 300 Daten
    $kfw300_rates = [];
    
    echo "Anzahl Programme in API: " . count($kfw_programs) . "\n";
    
    // Debug: Zeige erstes Entry
    if (count($kfw_programs) > 0) {
        echo "Beispiel-Entry (erstes Element):\n";
        echo "  Keys: " . implode(', ', array_keys($kfw_programs[0])) . "\n";
        if (isset($kfw_programs[0]['programm'])) {
            echo "  programm: " . $kfw_programs[0]['programm'] . " (Typ: " . gettype($kfw_programs[0]['programm']) . ")\n";
        }
    }
    
    // Zähle KfW 300 Einträge
    $kfw300_count = 0;
    foreach ($kfw_programs as $entry) {
        if (isset($entry['programm'])) {
            // Vergleiche als String UND als Integer
            if ($entry['programm'] == 300 || $entry['programm'] == '300') {
                $kfw300_count++;
                echo "KfW 300 Entry #$kfw300_count gefunden:\n";
                echo json_encode($entry, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE) . "\n";
            }
        }
    }
    echo "Gesamt KfW 300 Einträge gefunden: $kfw300_count\n";
    
    foreach ($kfw_programs as $entry) {
        if (isset($entry['programm']) && ($entry['programm'] == 300 || $entry['programm'] == '300')) {
            $laufzeit = isset($entry['laufzeit']) ? $entry['laufzeit'] : null;
            $sollzins = isset($entry['sollzins']) ? $entry['sollzins'] : null;
            
            if ($laufzeit === null || $sollzins === null) {
                echo "⚠ Entry übersprungen (fehlende Felder): laufzeit=$laufzeit, sollzins=$sollzins\n";
                continue;
            }
            
            // Mapping: laufzeit 1 = 4-10 Jahre, laufzeit 2 = 11-25 Jahre, laufzeit 3 = 26-35 Jahre
            if ($laufzeit == 1) {
                $kfw300_rates['4-10_jahre'] = ['sollzins' => $sollzins];
                echo "✓ API: 4-10 Jahre = $sollzins%\n";
            } elseif ($laufzeit == 2) {
                $kfw300_rates['11-25_jahre'] = ['sollzins' => $sollzins];
                echo "✓ API: 11-25 Jahre = $sollzins%\n";
            } elseif ($laufzeit == 3) {
                $kfw300_rates['26-35_jahre'] = ['sollzins' => $sollzins];
                echo "✓ API: 26-35 Jahre = $sollzins%\n";
            }
        }
    }
    
    // Validierung: Alle drei Laufzeiten gefunden?
    if (count($kfw300_rates) == 3) {
        echo "✓ KfW 300 Zinsen erfolgreich von API abgerufen\n";
        return $kfw300_rates;
    }
    
    echo "⚠ Nicht alle KfW 300 Laufzeiten in API gefunden (" . count($kfw300_rates) . "/3)\n";
    return null;
}

/**
 * KfW-Zinssätze scrapen (HTML Fallback)
 */
function scrape_kfw_zinsen() {
    echo "Scrape KfW-Zinsen (HTML Fallback)...\n";
    
    $url = 'https://www.kfw.de/inlandsfoerderung/Privatpersonen/Neubau/F%C3%B6rderprodukte/Wohneigentum-f%C3%BCr-Familien-(300)/';
    $html = fetch_html($url);
    
    if (!$html) {
        echo "Fehler beim Laden der KfW-Seite\n";
        return null;
    }
    
    $rates = [];
    
    // DOM-Parsing - Suche nach span-Elementen mit data-Attributen
    $dom = new DOMDocument();
    @$dom->loadHTML($html);
    $xpath = new DOMXPath($dom);
    
    // Suche nach span-Elementen mit KfW 300 Zinsdaten
    // Die Struktur ist: <span data-program-numbers="300" data-laufzeit="XX" data-interest-rate-type="sollzins">
    $spans = $xpath->query("//span[@data-program-numbers='300'][@data-interest-rate-type='sollzins']");
    
    echo "Gefundene KfW 300 Sollzins-Elemente: " . $spans->length . "\n";
    
    foreach ($spans as $span) {
        $laufzeit = $span->getAttribute('data-laufzeit');
        
        // Debug: Zeige alle Attribute
        echo "Debug: data-laufzeit='$laufzeit'";
        
        // Finde den inneren span mit class="link-labeling" der den Zinssatz enthält
        $label_spans = $xpath->query(".//span[@class='link-labeling']", $span);
        
        echo ", label_spans gefunden: " . $label_spans->length;
        
        if ($label_spans->length > 0) {
            $zins_text = trim($label_spans->item(0)->textContent);
            echo ", Text: '$zins_text'";
            
            // Parse "0,01 %" oder "1,09 %"
            if (preg_match('/(\d+[,\.]\d+)\s*%/', $zins_text, $match)) {
                $sollzins = floatval(str_replace(',', '.', $match[1]));
                echo ", Parsed: $sollzins%";
                
                // Mapping basierend auf data-laufzeit Attribut
                if ($laufzeit == '10') {
                    $rates['4-10_jahre'] = ['sollzins' => $sollzins];
                    echo " → ✓ 4-10 Jahre\n";
                }
                elseif ($laufzeit == '25') {
                    $rates['11-25_jahre'] = ['sollzins' => $sollzins];
                    echo " → ✓ 11-25 Jahre\n";
                }
                elseif ($laufzeit == '35') {
                    $rates['26-35_jahre'] = ['sollzins' => $sollzins];
                    echo " → ✓ 26-35 Jahre\n";
                }
                else {
                    echo " → ⚠ Laufzeit '$laufzeit' nicht erkannt\n";
                }
            } else {
                echo " → ⚠ Regex match fehlgeschlagen\n";
            }
        } else {
            echo " → ⚠ Keine link-labeling spans gefunden\n";
        }
    }
    
    if (count($rates) >= 3) {
        // Prüfe ob alle Werte gleich sind (dann stimmt was nicht)
        $values = array_column($rates, 'sollzins');
        $unique_values = array_unique($values);
        
        if (count($unique_values) === 1) {
            echo "⚠ Alle Zinssätze sind gleich ($values[0]%) - das ist unrealistisch!\n";
            echo "Verwerfe Ergebnis und nutze Fallback.\n";
            return null;
        }
        
        echo "✓ KfW 300 Zinsen erfolgreich gescraped: " . count($rates) . " Laufzeiten\n";
        return $rates;
    }
    
    echo "⚠ Nicht alle KfW-Zinsen gefunden (" . count($rates) . "/3)\n";
    return null;
}

/**
 * Bauzinsen von Interhyp scrapen
 */
function scrape_interhyp_zinsen() {
    echo "Scrape Interhyp-Zinsen...\n";
    
    $url = 'https://www.interhyp.de/ratgeber/was-muss-ich-wissen/zinsen-konditionen/interhyp-bauzinsen-charts.html';
    $html = fetch_html($url);
    
    if ($html) {
        $result = parse_bauzinsen($html, 'Interhyp');
        if ($result) return $result;
    }
    
    echo "⚠ Interhyp fehlgeschlagen\n";
    return null;
}

/**
 * Bauzinsen von Check24 scrapen
 */
function scrape_check24_zinsen() {
    echo "Scrape Check24-Zinsen...\n";
    
    $url = 'https://www.check24.de/baufinanzierung/zinsen/';
    $html = fetch_html($url);
    
    if ($html) {
        $result = parse_bauzinsen($html, 'Check24');
        if ($result) return $result;
    }
    
    echo "⚠ Check24 fehlgeschlagen\n";
    return null;
}

/**
 * Bauzinsen von Dr. Klein scrapen
 */
function scrape_drklein_zinsen() {
    echo "Scrape Dr. Klein-Zinsen...\n";
    
    $url = 'https://www.drklein.de/baufinanzierung-zinsen.html';
    $html = fetch_html($url);
    
    if ($html) {
        $result = parse_bauzinsen($html, 'Dr. Klein');
        if ($result) return $result;
    }
    
    echo "⚠ Dr. Klein fehlgeschlagen\n";
    return null;
}

/**
 * Intelligente Bauzinsen-Erkennung für alle Quellen
 */
function parse_bauzinsen($html, $source) {
    // Methode 1: Regex-Suche nach realistischen Bauzinsen (2% - 6%)
    preg_match_all('/(\d+[,.]?\d*)\s*%/u', $html, $all_matches);
    
    $found_rates = [];
    foreach ($all_matches[1] as $match) {
        $rate = floatval(str_replace(',', '.', $match));
        if ($rate >= 2.0 && $rate <= 6.0) {
            $found_rates[] = $rate;
        }
    }
    
    // Filtere unrealistische Werte (z.B. 100%, Jahreszahlen)
    $found_rates = array_filter($found_rates, function($rate) {
        return $rate < 10; // Nur Zinssätze unter 10%
    });
    
    if (count($found_rates) >= 3) {
        // Nehme den Median der gefundenen Zinssätze
        sort($found_rates);
        $median_index = floor(count($found_rates) / 2);
        $zins = $found_rates[$median_index];
        
        echo "Gefundene Zinssätze ($source): " . implode(', ', array_unique($found_rates)) . "%\n";
        echo "✓ Verwende Median: $zins%\n";
        
        return [
            'zinsbindung_10' => ['zins' => $zins]
        ];
    }
    
    // Methode 2: DOM-Parsing nach Tabellen
    $dom = new DOMDocument();
    @$dom->loadHTML($html);
    $xpath = new DOMXPath($dom);
    
    $rows = $xpath->query("//table//tr | //tbody//tr");
    
    foreach ($rows as $row) {
        $cells = $xpath->query(".//td | .//th", $row);
        if ($cells->length < 2) continue;
        
        $row_text = '';
        foreach ($cells as $cell) {
            $row_text .= ' ' . trim($cell->textContent);
        }
        
        // Suche nach "10 Jahre" Zeile
        if (preg_match('/10.*Jahre|Jahre.*10/iu', $row_text)) {
            foreach ($cells as $cell) {
                $text = trim($cell->textContent);
                if (preg_match('/(\d+[,.]?\d*)\s*%/', $text, $matches)) {
                    $zins = floatval(str_replace(',', '.', $matches[1]));
                    if ($zins >= 2 && $zins <= 6) {
                        echo "✓ $source-Zins 10 Jahre gefunden: $zins%\n";
                        return [
                            'zinsbindung_10' => ['zins' => $zins]
                        ];
                    }
                }
            }
        }
    }
    
    return null;
}

/**
 * Markt-Bauzinsen von mehreren Quellen scrapen
 */
function scrape_markt_zinsen() {
    echo "\n=== Markt-Bauzinsen (10 Jahre Zinsbindung) ===\n";
    
    // Versuche Interhyp
    $result = scrape_interhyp_zinsen();
    if ($result) return $result;
    
    // Fallback: Check24
    $result = scrape_check24_zinsen();
    if ($result) return $result;
    
    // Fallback: Dr. Klein
    $result = scrape_drklein_zinsen();
    if ($result) return $result;
    
    echo "⚠ Alle Markt-Zinsquellen fehlgeschlagen\n";
    return null;
}

/**
 * Alle Daten scrapen und JSON speichern
 */
function scrape_all() {
    echo "=== Start Scraping ===\n";
    echo date('Y-m-d H:i:s') . "\n\n";
    
    $json_file = __DIR__ . '/zinsen-kfw300.json';
    
    // Lade vorhandene Daten als Fallback
    $existing_data = null;
    if (file_exists($json_file)) {
        $existing_json = file_get_contents($json_file);
        $existing_data = json_decode($existing_json, true);
        echo "✓ Vorhandene JSON gefunden (Fallback)\n\n";
    }
    
    // KfW-Daten: Versuche zuerst die API
    echo "=== KfW 300 Zinsen ===\n";
    $kfw_rates = scrape_kfw_zinsen_api();
    $kfw_quelle = 'API';
    
    // Fallback 1: HTML-Scraping
    if (!$kfw_rates) {
        echo "→ API fehlgeschlagen, versuche HTML-Scraping...\n";
        $kfw_rates = scrape_kfw_zinsen();
        $kfw_quelle = 'HTML-Scraping';
    }
    
    // Fallback 2: Vorhandene JSON-Daten
    if (!$kfw_rates && $existing_data && isset($existing_data['kfw']['rates'])) {
        echo "→ HTML-Scraping fehlgeschlagen, verwende vorhandene JSON-Daten\n";
        $kfw_rates = $existing_data['kfw']['rates'];
        $kfw_quelle = 'Vorhandene JSON';
    }
    
    // Fallback 3: Hart-codierte aktuelle Werte
    if (!$kfw_rates) {
        echo "→ Verwende hart-codierte Fallback KfW 300 Daten (aktuelle Zinsen)\n";
        $kfw_rates = [
            '4-10_jahre' => ['sollzins' => 0.01],
            '11-25_jahre' => ['sollzins' => 1.09],
            '26-35_jahre' => ['sollzins' => 1.32]
        ];
        $kfw_quelle = 'Hart-codiert';
    }
    
    echo "KfW-Quelle: $kfw_quelle\n\n";
    
    // Markt-Bauzinsen
    $interhyp_rates = scrape_markt_zinsen();
    
    if (!$interhyp_rates) {
        if ($existing_data && isset($existing_data['interhyp']['rates'])) {
            echo "→ Verwende vorhandene Interhyp-Daten aus JSON\n";
            $interhyp_rates = $existing_data['interhyp']['rates'];
        } else {
            echo "→ Verwende hart-codierte Fallback Interhyp-Daten\n";
            $interhyp_rates = [
                'zinsbindung_10' => ['zins' => 3.96]
            ];
        }
    }
    
    // JSON-Struktur erstellen
    $data = [
        'kfw' => [
            'rates' => $kfw_rates,
            'quelle' => $kfw_quelle
        ],
        'interhyp' => [
            'rates' => $interhyp_rates
        ],
        'updated' => date('c') // ISO 8601 Format
    ];
    
    // JSON speichern
    $json_content = json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    
    if (file_put_contents($json_file, $json_content)) {
        echo "\n✓ JSON erfolgreich gespeichert: $json_file\n";
        echo "Größe: " . filesize($json_file) . " Bytes\n";
        echo "\nInhalt:\n";
        echo "KfW-Quelle: $kfw_quelle\n";
        echo "KfW 4-10 Jahre: " . $kfw_rates['4-10_jahre']['sollzins'] . "%\n";
        echo "KfW 11-25 Jahre: " . $kfw_rates['11-25_jahre']['sollzins'] . "%\n";
        echo "KfW 26-35 Jahre: " . $kfw_rates['26-35_jahre']['sollzins'] . "%\n";
        echo "Interhyp 10 Jahre: " . $interhyp_rates['zinsbindung_10']['zins'] . "%\n";
        return true;
    } else {
        echo "\n✗ Fehler beim Speichern der JSON-Datei\n";
        return false;
    }
}

// Script ausführen
scrape_all();

echo "\n=== Scraping abgeschlossen ===\n";
?>

File: scrape_kfw_zinsen.py
#!/usr/bin/env python3
"""
KfW 300 Zinssätze Scraper mit Interhyp-Vergleich
Lädt täglich die aktuellen Zinssätze von der KfW-Website und Interhyp und speichert sie als JSON
"""

import requests
from bs4 import BeautifulSoup
import json
from datetime import datetime
import re

def scrape_kfw_zinsen():
    """
    Scrapt die KfW-Website für Programm 300 Zinssätze
    """
    # KfW 300 Konditionsseite
    url = "https://www.kfw.de/inlandsfoerderung/Privatpersonen/Neubau/F%C3%B6rderprodukte/Wohneigentum-f%C3%BCr-Familien-(300)/"
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Zinssätze extrahieren (anpassen an tatsächliche HTML-Struktur)
        # Dies ist ein Template - muss an die tatsächliche Struktur angepasst werden
        zinsen = {
            "updated": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            "program": "KfW 300 - Wohneigentum für Familien",
            "rates": {
                "4-10_jahre": {
                    "laufzeit": "4 bis 10 Jahre",
                    "zinsbindung": "10 Jahre",
                    "tilgungsfrei": "1 bis 2 Jahre",
                    "sollzins": 0.0,
                    "effektivzins": 0.0
                },
                "11-25_jahre": {
                    "laufzeit": "11 bis 25 Jahre",
                    "zinsbindung": "10 Jahre",
                    "tilgungsfrei": "1 bis 3 Jahre",
                    "sollzins": 0.0,
                    "effektivzins": 0.0
                },
                "26-35_jahre": {
                    "laufzeit": "26 bis 35 Jahre",
                    "zinsbindung": "10 Jahre",
                    "tilgungsfrei": "1 bis 5 Jahre",
                    "sollzins": 0.0,
                    "effektivzins": 0.0
                }
            },
            "source_url": url
        }
        
        # Versuche Zinssätze zu finden (diese Selektoren müssen angepasst werden!)
        # Beispiel für mögliche Struktur:
        zins_tabelle = soup.find('table', {'class': lambda x: x and 'konditionen' in x.lower() if x else False})
        
        if zins_tabelle:
            rows = zins_tabelle.find_all('tr')
            for row in rows:
                cols = row.find_all('td')
                if len(cols) >= 2:
                    text = ' '.join([col.get_text(strip=True) for col in cols])
                    
                    # Suche nach Zinssätzen (Format: X,XX %)
                    zins_matches = re.findall(r'(\d+[,\.]\d+)\s*%', text)
                    
                    if '4' in text and '10' in text and zins_matches:
                        zinsen['rates']['4-10_jahre']['sollzins'] = float(zins_matches[0].replace(',', '.'))
                        if len(zins_matches) > 1:
                            zinsen['rates']['4-10_jahre']['effektivzins'] = float(zins_matches[1].replace(',', '.'))
                    
                    elif '11' in text and '25' in text and zins_matches:
                        zinsen['rates']['11-25_jahre']['sollzins'] = float(zins_matches[0].replace(',', '.'))
                        if len(zins_matches) > 1:
                            zinsen['rates']['11-25_jahre']['effektivzins'] = float(zins_matches[1].replace(',', '.'))
                    
                    elif '26' in text and '35' in text and zins_matches:
                        zinsen['rates']['26-35_jahre']['sollzins'] = float(zins_matches[0].replace(',', '.'))
                        if len(zins_matches) > 1:
                            zinsen['rates']['26-35_jahre']['effektivzins'] = float(zins_matches[1].replace(',', '.'))
        
        # Fallback: Wenn Scraping fehlschlägt, verwende Platzhalter
        if all(zinsen['rates'][key]['sollzins'] == 0.0 for key in zinsen['rates']):
            print("⚠️ Warnung: Keine Zinssätze gefunden. Bitte HTML-Struktur prüfen!")
            print("Verwende Platzhalter-Werte...")
            
            # Platzhalter (aktualisiert: 11.02.2026)
            zinsen['rates']['4-10_jahre']['sollzins'] = 0.01
            zinsen['rates']['4-10_jahre']['effektivzins'] = 0.01
            zinsen['rates']['11-25_jahre']['sollzins'] = 1.09
            zinsen['rates']['11-25_jahre']['effektivzins'] = 1.10
            zinsen['rates']['26-35_jahre']['sollzins'] = 1.32
            zinsen['rates']['26-35_jahre']['effektivzins'] = 1.33
            zinsen['manual_update_required'] = True
        
        print("✅ KfW Zinssätze extrahiert")
        
        return zinsen
        
    except Exception as e:
        print(f"❌ Fehler beim KfW-Scraping: {e}")
        
        # Erstelle minimale Daten mit Fehlermeldung
        error_data = {
            "updated": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            "error": str(e),
            "message": "Scraping fehlgeschlagen. Bitte manuell prüfen.",
            "program": "KfW 300 - Wohneigentum für Familien",
            "rates": {
                "4-10_jahre": {
                    "laufzeit": "4 bis 10 Jahre",
                    "zinsbindung": "10 Jahre",
                    "tilgungsfrei": "1 bis 2 Jahre",
                    "sollzins": None,
                    "effektivzins": None
                },
                "11-25_jahre": {
                    "laufzeit": "11 bis 25 Jahre",
                    "zinsbindung": "10 Jahre",
                    "tilgungsfrei": "1 bis 3 Jahre",
                    "sollzins": None,
                    "effektivzins": None
                },
                "26-35_jahre": {
                    "laufzeit": "26 bis 35 Jahre",
                    "zinsbindung": "10 Jahre",
                    "tilgungsfrei": "1 bis 5 Jahre",
                    "sollzins": None,
                    "effektivzins": None
                }
            },
            "source_url": "https://www.kfw.de/inlandsfoerderung/Privatpersonen/Neubau/F%C3%B6rderprodukte/Wohneigentum-f%C3%BCr-Familien-(300)/"
        }
        
        return error_data

def scrape_interhyp_zinsen():
    """
    Scrapt Interhyp-Zinssätze für Zinsbindung 10 Jahre, Beleihungsauslauf <90%
    """
    url = "https://www.interhyp.de/zinsen/"
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.content, 'html.parser')
        
        interhyp_data = {
            "zinsbindung_10": {
                "zinsbindung": "10 Jahre",
                "beleihung": "< 90%",
                "zins": None
            }
        }
        
        # Suche nach der Zinssatz-Tabelle
        # Interhyp hat oft eine Tabelle mit Zinsbindung als Zeilen und Beleihung als Spalten
        
        # Strategie 1: Suche nach Tabelle mit "Zinsbindung" und "Beleihungsauslauf"
        tables = soup.find_all('table')
        
        for table in tables:
            # Suche nach Header mit "Beleihungsauslauf >90" oder ähnlich
            headers = table.find_all(['th', 'td'])
            header_texts = [h.get_text(strip=True) for h in headers]
            
            # Prüfe ob es eine Zinsbindungs-Tabelle ist
            if any('Beleihung' in text or '90' in text for text in header_texts):
                print("📊 Gefundene Tabelle mit Beleihung-Spalten")
                
                # Finde alle Zeilen
                rows = table.find_all('tr')
                
                for row in rows:
                    cells = row.find_all(['td', 'th'])
                    if len(cells) >= 4:  # Mindestens 4 Spalten erwartet
                        row_text = cells[0].get_text(strip=True)
                        
                        # Suche nach Zeile mit "10" für 10 Jahre Zinsbindung
                        if row_text == '10' or '10 Jahre' in row_text:
                            print(f"✓ Gefundene Zeile: 10 Jahre Zinsbindung")
                            
                            # Die Spalten sind typisch: [Zinsbindung, <70, =80, >90]
                            # Für <90% nehmen wir die letzte Spalte (>90 ist außerhalb)
                            # oder die vorletzte Spalte (wenn =80 gemeint ist)
                            
                            # Versuche den Zins aus der letzten Spalte zu extrahieren
                            zins_text = cells[-1].get_text(strip=True)
                            
                            # Extrahiere Zahl im Format X,XX % oder X.XX %
                            zins_match = re.search(r'(\d+[,\.]\d{1,2})\s*%?', zins_text)
                            
                            if zins_match:
                                zins_str = zins_match.group(1).replace(',', '.')
                                interhyp_data['zinsbindung_10']['zins'] = float(zins_str)
                                print(f"✅ Interhyp Zins (10 Jahre, >90%): {zins_str}%")
                                break
                
                if interhyp_data['zinsbindung_10']['zins'] is not None:
                    break
        
        # Strategie 2: Fallback - Suche nach spezifischen Mustern im Text
        if interhyp_data['zinsbindung_10']['zins'] is None:
            print("⚠️ Tabelle nicht gefunden, versuche Textsuche...")
            
            # Suche nach Muster wie "10" gefolgt von Zinssätzen
            text_content = soup.get_text()
            lines = text_content.split('\n')
            
            for i, line in enumerate(lines):
                if line.strip() == '10' or 'Zinsbindung Tranche' in line:
                    # Schaue in den nächsten Zeilen nach Zinssätzen
                    for j in range(i, min(i+5, len(lines))):
                        zins_matches = re.findall(r'(\d+[,\.]\d{2})\s*%', lines[j])
                        if zins_matches:
                            # Nehme den letzten gefundenen Wert (oft >90%)
                            zins_str = zins_matches[-1].replace(',', '.')
                            interhyp_data['zinsbindung_10']['zins'] = float(zins_str)
                            print(f"✅ Interhyp Zins aus Text: {zins_str}%")
                            break
                    
                    if interhyp_data['zinsbindung_10']['zins'] is not None:
                        break
        
        # Wenn nichts gefunden wurde, Platzhalter verwenden
        if interhyp_data['zinsbindung_10']['zins'] is None:
            print("⚠️ Warnung: Keine Interhyp-Zinssätze gefunden. Verwende Platzhalter...")
            interhyp_data['zinsbindung_10']['zins'] = 3.96  # Platzhalter aktualisiert
        
        return interhyp_data
        
    except Exception as e:
        print(f"⚠️ Fehler beim Interhyp-Scraping: {e}")
        print("Verwende Platzhalter-Wert...")
        return {
            "zinsbindung_10": {
                "zinsbindung": "10 Jahre",
                "beleihung": "< 90%",
                "zins": 3.96  # Platzhalter aktualisiert
            }
        }

def scrape_all():
    """
    Scrapt KfW und Interhyp Zinssätze und kombiniert sie
    """
    print("🔄 Starte Scraping...")
    print("\n--- KfW 300 ---")
    kfw_data = scrape_kfw_zinsen()
    
    print("\n--- Interhyp ---")
    interhyp_data = scrape_interhyp_zinsen()
    
    # Kombiniere beide Datenquellen
    combined_data = {
        "updated": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "kfw": {
            "program": kfw_data.get('program', 'KfW 300'),
            "rates": kfw_data.get('rates', {}),
            "source_url": kfw_data.get('source_url', '')
        },
        "interhyp": {
            "rates": interhyp_data,
            "source_url": "https://www.interhyp.de/zinsen/"
        }
    }
    
    # Speichere kombinierte Daten
    with open('zinsen-kfw300.json', 'w', encoding='utf-8') as f:
        json.dump(combined_data, f, indent=2, ensure_ascii=False)
    
    print("\n✅ Alle Daten erfolgreich gespeichert!")
    print(json.dumps(combined_data, indent=2, ensure_ascii=False))
    
    return combined_data

if __name__ == "__main__":
    scrape_all()
